### PR TITLE
[Hashes] Adds a comparator and changes the initializers a bit

### PR DESCRIPTION
**Pull request by zamith**
Adds the IndifferentAccessComparator, that allows to also use symbols as
keys, even when the Hash was created with Strings as keys.

The default value in the actual initialize method was not being used at
all (probably a left over), so I remove it. I also added a few other
ways to initialize hashes, so that you most likely will be able to
initialize it any way you want without having to pass nils, just to
comply with the arity.

### CODE REVIEW

File: src/hash.cr [hunk 1]
```diff
@@ -27,21 +27,43 @@ class Hash(K, V)
     end
   end
 
+  module IndifferentAccessComparator
+    def self.hash(sym : Symbol)
+      sym.to_s.downcase.hash
```
Comment (asterite) [inline src/hash.cr]: Why the downcase?

**Comment by asterite** [comment]
Thanks!

I'll keep the pull request on hold because I'm not sure I like the current way we do things in Hash. There are many downsides to the current approach:

1. The `@comp` is a union of all the comparators used to initialize hashes. That makes Hash (a bit) slower. The correct way would be to make Hash be something like Hash(K, V, C), where C is the comparator class. But I don't know of any language that implements a Hash like this. And then we would maybe need default type variables, so Hash would be Hash(K, V, C = StandardCompartor), but that menas adding a feature to the language...
2. If you have, say, a hash with indifferent access, if you do `h[:foo] = 1; h["bar"] = 2`, then `h.keys` will be `[:foo, "bar"]`, that is, a type mix, because the first one that didn't exist enters the hash. In Ruby HashWithIndifferentAccess always returns strings as keys.

We originally did the comparator thing to implement HTTP headers because they are case insensitive, but I finally implemented them as a [struct that wraps a Hash](https://github.com/manastech/crystal/blob/master/src/http/headers.cr). It wasn't hard, and all keys are capitalized when asked. And I'm not sure there would be many variants in the comparator (case insensitive, string/symbol, what else?). Or we could have a HashWithComparator(K, V, C) that wraps (or extends) Hash and adds a comparator (or maybe a transformer would be better).

I'm just explaining all of this because I don't think it's worth it to add more comparators to Hash if we still don't know if this design will survive.

/cc @waj

**Comment by jhass** [comment]
How about swapping the delegation around? Get rid of the comparator and have a `HashWithCaseinsensitiveAccess` (okay, needs a shorter name :P) that delegates to a normal Hash and normalizes all inputs and outputs.

File: src/hash.cr (hunk 1 above)
Comment (zamith) [inline src/hash.cr]: For no reason really. I will remove it.

**Comment by zamith** [comment]
@asterite I have no problem removing the comparator, but what about the changes to the initializers? The main reason for them is so that you can initialize a hash in your preferred way, without having to pass nils. 

@jhass That is a possibility, but I would not do that in the language. I'd rather have that done by the user/lib that needs it, and remove the comparators.

**Comment by asterite** [comment]
@jhass I agree with you (but as @zamith says, I'm not sure I would include that in the standard library). Also, if HashWithCaseInsensitiveAccess is a struct that wraps a Hash, there's no performance penalty for the delegation :-)

@zamith Yes, the refactor of the initializers is good indeed :-)

**Comment by zamith** [comment]
@asterite Ok, removed the comparator.

**Comment by asterite** [comment]
Thanks!!

**Pull request merged**